include/nuttx/sensors/apds9922.h

```diff
@@ -0,0 +1,315 @@
+/****************************************************************************
+ * include/nuttx/sensors/apds9922.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_SENSORS_APDS9922_H
+#define __INCLUDE_NUTTX_SENSORS_APDS9922_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/irq.h>
+#include <nuttx/i2c/i2c_master.h>
+#include <nuttx/sensors/ioctl.h>
+
+#if defined(CONFIG_SENSORS_APDS9922)
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define APDS9922_I2C_ADDR   (0x53)
+#define APDS9922_ID_VAL     (0xb3)
+
+#define PS_DEF_THRESHU      (0x07ff)
+#define PS_DEF_THRESHL      (0)
+#define PS_DEF_CANCEL_LVL   (0)
+#define PS_DEF_PULSES       (0x08)
+#define ALS_DEF_THRESHU     (0x0fffff)
+#define ALS_DEF_THRESHL     (0)
+#define PS_DEF_PERSISTANCE  (0)
+#define ALS_DEF_PERSISTANCE (0)
+#define ALS_DEF_VAR         (0)
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+enum apds9922_als_rate
+{
+  ALS_RATE25MS = 0,
+  ALS_RATE50MS,
+  ALS_RATE100MS,
+  ALS_RATE200MS,
+  ALS_RATE500MS,
+  ALS_RATE1000MS,
+  ALS_RATE2000MS,
+  ALS_RATE4000MS,
+};
+
+static const uint32_t apds9922_als_res_table[] =
```

Comment:
This is not a `Public Types`. Can it be moved to C-file?

Reply:
I have a test app that makes use of these so i included them in case others wanted to, so I'll move as the test app is not worthy of a PR. Or I'll make the test app better :)

Reply:
@pkarashchenko Before I remove these, the use of enums (rather than just accepting any int value) mean, in theory, only valid values can be used when using the public setup structs. At least the valid values are then clear to everyone in the header.

Is it not better to have the valid values sated like this (even though I do check for valid values in the relevant functions)?

Perhaps this is just personal preference; but if it is wrong (by NuttX or C89 rules, or whatever) I will move them of course.

Please advise.

Reply:
The values are actually sequential integers and map to the register value very conveniently, and the names are descriptive to make it clear what the integer value represents in milliseconds.

Isn't this unnecessary? What are you trying to achieve - if I understand that I can come up with something, hopefully?

Reply:
Each source file that will include the current version of the header will have its own copy static array data even if it might not use it. In case of linker garbage collector or link time optimization it will be wiped of course, but without such optimization the image size might be increasing without real value.
So my main goal is not to have static arrays in header files.

Reply:
> Each source file that will include the current version of the header will have its own copy static array data even if it might not use it. In case of linker garbage collector or link time optimization it will be wiped of course, but without such optimization the image size might be increasing without real value. So my main goal is not to have static arrays in header files.

Thanks for explaining - I will rethink this now I understand the reason :)

Reply:
I am pushing a commit that uses your first suggestion. Not as "neat" as my original solution but I understand the reason and hope this is acceptable.

I won't squash yet in case it's a cat's breakfast and I need to revert it.

Reply:
squashed